src/ekat/ekat_pack.hpp

```diff
@@ -314,6 +314,22 @@ OnlyPackReturn<PackType, typename PackType::scalar> max (const PackType& p) {
   return v;
 }
 
+// return a+p[0]+p[1]+...
+template <bool Serialize, typename PackType, typename Scalar> KOKKOS_INLINE_FUNCTION
+OnlyPackReturn<PackType, typename PackType::scalar> reduce_sum (const PackType& p, const Scalar a=Scalar()) {
+  Scalar return_val = Scalar(a);
+  if (Serialize) {
+    for (int i = 0; i < PackType::n; ++i) {
+      return_val += p[i];
+    }
+  } else {
+    typename PackType::scalar sum = p[0];
```

Comment:
Probably when we look at the best way to do this eventually, we'll find that sum should be init'ed to 0 and then the loop should be over [0,n).

Reply:
Since `a` is passed by value, we could even recycle `a` as the sum, so we avoid creating an extra tmp.

Maybe, the best impl would be to have two routines:

```
void reduce_sum(const Pack& p, Scalar& sum);

void reduce_sum(const Pack& p);
```
The latter would create a temp for the sum, and then call the former, while the former would sum straight into the sum input.

Reply:
Also, I think you are treating the result differently in the two cases: in the first, the type of  `return_val` is the input template parameter, while in the second is the scalar type of the pack. This could cause different behaviors in the two cases, when `Scalar!=PackType::scalar`.

Perhaps it's best to remove the Scalar template, and deducing it from the PackType?

Reply:
I implemented the two versions of `reduce_sum` you suggested and removed the Scalar template.